server/cmwell-data-tools/src/main/scala/cmwell/tools/data/downloader/consumer/BufferFillerActor.scala

```diff
@@ -142,13 +144,20 @@ class BufferFillerActor(threshold: Int,
       tsvCounter += 1
 
     case GetData if buf.nonEmpty =>
-      sender ! buf.dequeue()
+     sender ! ( buf.dequeue() match {
```

Comment:
deque will never return None (it will throw NoSuchElementException that is the reason for the below `if buf.nonEmpty` check.

Reply:
It is only there as the Downloader.scala has a check for None after it receives the message from GetData... seems to be None is used to signal end of the stream... Maybe it is redundant?

Reply:
Well it's my mistake. The dequeue method will never return null but the element itself can be.  I looked in the code and found the following:
https://github.com/thomsonreuters/CM-Well/blob/f7da7b700e619f42c236f201d281cf100285a933/server/cmwell-data-tools/src/main/scala/cmwell/tools/data/downloader/consumer/BufferFillerActor.scala#L229
which will cause the element to be None for the Downloader.
This "signal" is more "smart" than the horizon because it is set only when the buffer is empty (e.g. the down stream already took all the elements). It can be used to notify of horizon+all TSVs were taken by the down stream.
I never knew about it, it can maybe improve the horizon feature but I will not insist on it.